[fix #166] REST endpoint for Subhub to populate Salesforce

jpetto opened the pull request:
Create a simple API to capture Stripe transactions and put them in SFDC/SFMC.

All data will come to a single view/endpoint. That view will check the API token, then route the payload to the appropriate task based on the `event_type` sent in the payload. Tasks perform various bits of data transformation and logic before inserting/updating records in Salesforce.

Fixes #166

jpetto commented:
@pmac Still very much in draft mode, but would love a sanity check when you get a chance. 😄

pmac commented on basket/news/views.py:
```diff
@@ -939,3 +955,36 @@ def update_user_task(request, api_call_type, data=None, optin=False, sync=False)
+            return HttpResponseJSON({'status': 'ok'})
+        else:
+            # TODO: set a better HTTP response code? 409?
+            return HttpResponseJSON({'status': 'unknown event type'})
```
At least return a 400 and a status of `error` and this as `desc`. This is good though 👍

pmac commented on basket/news/views.py:
```diff
@@ -939,3 +955,36 @@ def update_user_task(request, api_call_type, data=None, optin=False, sync=False)
+
+    try:
+        data = json.loads(request.body)
+    except ValueError as _:
```
If you don't need the exception object just leave the `as _` off.
```suggestion
    except ValueError:
```

pmac commented on basket/news/views.py:
```diff
@@ -939,3 +955,36 @@ def update_user_task(request, api_call_type, data=None, optin=False, sync=False)
+        data = json.loads(request.body)
+    except ValueError as _:
+        statsd.incr('subhub_post.message.json_error')
+        sentry_client.captureException(data={'extra': {'request.body': request.body}})
```
👍

pmac commented on basket/news/utils.py:
```diff
@@ -280,9 +280,10 @@ def language_code_is_valid(code):
 
 
-def get_user_data(token=None, email=None, extra_fields=None, get_fxa=False):
+def get_user_data(token=None, email=None, stripe_id=None, extra_fields=None, get_fxa=False):
```
Still working on making this work I guess?

jpetto replied:
Yeah - just put this in as a starting point.

pmac commented on basket/news/tasks.py:
```diff
@@ -717,6 +717,157 @@ def record_source_url(email, source_url, newsletter_id):
+        except sfapi.SalesforceMalformedRequest as e:
+            sfdc.opportunity.create(transaction_data)
+    else:
+        # TODO: do we need to do anything here if the user wasn't found?
```
Maybe there's enough info in the data to create the user if none is found? Really only need email address.

jpetto replied:
According to the doc, the only user identifiable info in these payloads is the Stripe ID, so I don't think we can do much?

jpetto commented on basket/news/tasks.py:
```diff
@@ -717,6 +717,157 @@ def record_source_url(email, source_url, newsletter_id):
+        try:
+            # will raise a SalesforceMalformedRequest if not found
+            sfdc.opportunity.update('PMT_Transaction_ID__c/{}'.format(charge_id), transaction_data)
+            # TODO: is this the proper syntax? unsure about the use/purpose of the first param here - is there some magic happening that makes this a key?
```
Would love some input on this bit. 😄

pmac replied:
It's the lookup basically. The first bit is the filed to look in, and the second is the data.

https://github.com/simple-salesforce/simple-salesforce#other-options

jpetto commented:
@glogiotatidis Do you have time to review this during your morning tomorrow? Tests are only stubs at the moment, but it would be great to get this up on dev for end to end testing of the Salesforce API.

glogiotatidis commented on basket/news/utils.py:
```diff
@@ -280,9 +280,10 @@ def language_code_is_valid(code):
 
 
-def get_user_data(token=None, email=None, extra_fields=None, get_fxa=False):
+def get_user_data(token=None, email=None, payee_id=None, extra_fields=None, get_fxa=False):
```
This should be fine but generally speaking I'm careful adding more parameters out of order, especially in a function with all optional parameters. I did a quick search and the only place where we don't explicitly name the parameter is in https://github.com/mozmeao/basket/blob/79624a385bcef27c639c312916282b02d11edac6/basket/news/tasks.py#L386, so maybe you can change that to `get_user_data(token=data.get('token'), email=data.get('email'), extra_fields=['id']))` just to be on the safe side.

jpetto replied:
Ah, nice catch!

I did think maybe `get_user_data` needs a refactor - maybe taking the (now) three identifiers and expecting a dict instead - something like `{ 'type': 'token', 'value': '123456' }`. Anyway, I'll leave that for later clean up.

glogiotatidis commented on basket/news/tasks.py:
```diff
@@ -717,6 +717,250 @@ def record_source_url(email, source_url, newsletter_id):
+        # just in case amount data is missing/malformed
+        try:
+            # amount values are cents - convert to dollars
+            amount_paid = int(data['amount']) / 100
```
This always returns `int`. 
```python
int(108.5) / 100 = 1
```
Just a heads up, not sure if that's what you want here.

jpetto replied:
Hmmm, good catch. Will replace with:

`int(data['amount']) / float(100)`

That should do it, right?

glogiotatidis replied:
Yes that works. Another trick is

`int(data['amount']) / 100.0`

glogiotatidis commented on basket/news/tasks.py:
```diff
@@ -717,6 +717,250 @@ def record_source_url(email, source_url, newsletter_id):
+            # determine if this is an initial or recurring payment
+            try:
+                # look for existing opportunities with the current subscription_id
+                _ = sfdc.opportunity.get_by_custom_id('PMT_Subscription_ID__C', data['subscription_id'])
```
Why do you need `_`?

jpetto replied:
Meant to get rid of that - was just coding quickly and wasn't sure if I was going to be using the result. Will remove.

jpetto commented:
@glogiotatidis Made a commit addressing your feedback. If you have time to verify tomorrow morning, that would be great. Hopefully we can push this out to dev tomorrow. 😁

glogiotatidis commented:
Looks good, thanks!
